### PR TITLE
Fix ability to sort course students table by completed date.

### LINKED ISSUE #1969: Course Completed Header Not Sorting Correctly
### Reproduction Steps

+ HS-185936
+ Head into _LifterLMS->Reporting->Courses->Select a Course->Students_
+ From the clickable sorting headers click on Completed

### Expected Behavior

+ The students should be organized by date completed

### Actual Behavior

+ Not that.

https://www.loom.com/share/e09dc01cd16a4e9d8b0d08c61bd7ef82

### Error Messages / Logs

+ Include any relevant error messages or log files
```
<!-- Paste error logs / backtraces below this line -->

```

### System and Environment Information

<details>
<summary>System Report</summary>

```
Wordpress
-------------------------------------------

Home Url: https://natalie-test-site.myliftersite.com
Site Url: https://natalie-test-site.myliftersite.com
Login Url: https://natalie-test-site.myliftersite.com/wp-login.php
Version: 5.9
Debug Mode: No
Debug Log: No
Debug Display: Yes
Locale: en_US
Multisite: No
Page For Posts: Not Set
Page On Front: Not Set
Permalink Structure: /%postname%/
Show On Front: posts
Wp Cron: Yes


Settings
-------------------------------------------

Version: 5.8.0
Db Version: 5.8.0
Course Catalog: Course Catalog (#35) [https://natalie-test-site.myliftersite.com/courses/]
Membership Catalog: Membership Catalog (#36) [https://natalie-test-site.myliftersite.com/memberships/]
Student Dashboard: Dashboard (#38) [https://natalie-test-site.myliftersite.com/dashboard/]
Checkout Page: Checkout (#715) [https://natalie-test-site.myliftersite.com/checkout/]
Course Catalog Per Page: 9
Course Catalog Sorting: menu_order,ASC
Membership Catalog Per Page: 9
Membership Catalog Sorting: menu_order
Site Membership: Not Set
Courses Endpoint: my-courses
Edit Endpoint: edit-account
Lost Password Endpoint: lost-password
Vouchers Endpoint: redeem-voucher
Autogenerate Username: no
Password Strength Meter: no
Minimum Password Strength: 
Terms Required: yes
Terms Page: Activate (#443) [https://natalie-test-site.myliftersite.com/activate/]
Checkout Names: 
Checkout Address: 
Checkout Phone: 
Checkout Email Confirmation: no
Open Registration: yes
Registration Names: 
Registration Address: 
Registration Phone: 
Registration Voucher: 
Registration Email Confirmation: no
Account Names: 
Account Address: 
Account Phone: 
Account Email Confirmation: no
Confirmation Endpoint: confirm-payment
Force Ssl Checkout: no
Country: US
Currency: USD
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Decimals: 2
Trim Zero Decimals: no
Recurring Payments: yes
Email From Address: team+sandbox@lifterlms.com
Email From Name: A LifterLMS Sandbox
Email Footer Text: 
Email Header Image: 
Cert Bg Width: 1123
Cert Bg Height: 794
Cert Legacy Compat: no


Constants
-------------------------------------------

LLMS_REMOVE_ALL_DATA: undefined
LLMS_REST_DISABLE: undefined
LLMS_SITE_FEATURE_RECURRING_PAYMENTS: undefined
LLMS_SITE_IS_CLONE: undefined


Gateways
-------------------------------------------

Stripe: Enabled
Stripe Test Mode: Enabled
Stripe Logging: no
Stripe Order: 1
Manual: Enabled
Manual Logging: no
Manual Order: 1


Server
-------------------------------------------

Mysql Version: 5.7.37
Php Curl: Yes
Php Default Timezone: UTC
Php Fsockopen: Yes
Php Max Input Vars: 5000
Php Max Upload Size: 512 MB
Php Memory Limit: 256M
Php Post Max Size: 1024M
Php Soap: Yes
Php Suhosin: No
Php Time Limt: 30
Php Version: 8.0.15
Software: Apache/2.4.52 (Unix) OpenSSL/1.1.1
Wp Memory Limit: 256M


Browser
-------------------------------------------

HTTP USER AGENT: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36


Theme
-------------------------------------------

Name: Divi
Version: 4.14.7
Themeuri: http://www.elegantthemes.com/gallery/divi/
Authoruri: http://www.elegantthemes.com
Template: 
Child Theme: No
Llms Support: No


Plugins
-------------------------------------------

Beaver Builder Plugin (Lite Version): 2.5.1.3
LifterLMS: 5.8.0
LifterLMS Advanced Videos: 1.0.0-beta.17
LifterLMS Custom Fields: 2.0.2
LifterLMS Labs: 1.6.0
LifterLMS PDFs: 1.0.0
LifterLMS Stripe Payment Gateway: 5.4.0
Loco Translate: 2.5.8
User Switching: 1.5.8
WP Rollback: 1.7.1


Integrations
-------------------------------------------

BbPress: No
BuddyPress: No
Videos: Vimeo: No
Videos: Wistia: No
Videos: YouTube: No


Template Overrides
-------------------------------------------




```

</details>

This issue has be recreated:
+ [x] Locally
+ [ ] On a staging site
+ [x] On a production website
+ [] With only LifterLMS and a default theme

### Browser, Device, and Operating System Information

+ Browser name and version
+ Operating System name and version
+ Device name and version (if applicable)

### PR DESCRIPTION
**Pull request by pondermatic**
## Description

* Added ability to sort by 'completed' in `LLMS_Table_Course_Students::get_results()`.
* Added a subquery for completed date in `LLMS_Student_Query::sql_select()`.
* Added `$meta_key` argument in `LLMS_Student_Query::sql_subquery()`.
* Added unit tests on `LLMS_Table_Course_Students` based on `LLMS_Test_Table_Students`.

Fixes #1969.

## How has this been tested?

Manually and automatically.

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards.

### CODE REVIEW

File: includes/class.llms.student.query.php [hunk 1]
```diff
@@ -417,16 +419,18 @@ private function sql_status_in( $column = 'status' ) {
 	}
 
 	/**
-	 * Generate an SQL subquery for the dynamic status or date values in the main query
+	 * Generate an SQL subquery for the dynamic status or date values in the main query.
```
Comment (eri-trabiccolo) [inline includes/class.llms.student.query.php]: I would change this to reflect the new "ability" added :D

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]
Looks very solid to me.

File: includes/class.llms.student.query.php (hunk 1 above)
Comment (pondermatic) [reply includes/class.llms.student.query.php]: Is this good?
```Generate an SQL subquery for the meta key in the main query.```

Comment (eri-trabiccolo) [reply includes/class.llms.student.query.php]: I think so.

File: includes/class.llms.student.query.php [hunk 2]
```diff
@@ -417,16 +419,18 @@ private function sql_status_in( $column = 'status' ) {
...
 	 *
 	 * @since 3.13.0
+	 * @since [version] Add `$meta_key` argument.
 	 *
 	 * @param string $column Column name.
+	 * @param string $meta_key Optional meta key to use in the WHERE condition. Defaults to '_status'.
```
Comment (eri-trabiccolo) [inline includes/class.llms.student.query.php]:
```suggestion
	 * @param string $column   Column name.
	 * @param string $meta_key Optional meta key to use in the WHERE condition. Defaults to '_status'.
```

**Review by eri-trabiccolo — APPROVED** [review]
Looks good to me.

**Pull request merged**